apps/studio/src/components/ResourceSelector/useResourceStack.tsx

```diff
@@ -0,0 +1,209 @@
+      resourceId: String(existingResource?.id),
+      siteId: String(siteId),
+    })
+
+  const nestedChildrenOfExistingResource = useMemo(
+    (): ResourceItemContent[] =>
+      nestedChildrenOfExistingResourceResult?.items ?? [],
+    [nestedChildrenOfExistingResourceResult?.items],
+  )
+
+  const isResourceItemDisabled = useCallback(
+    (resourceItem: ResourceItemContent): boolean => {
+      // If there is no existing resource,
+      // Then we are linking the resource and not moving any resource
+      // Thus, no checks are needed because we can link to any resource
+      if (!existingResource) return false
+
+      // A resource should not be able to move to within itself
+      if (existingResource.id === resourceItem.id) return true
+
+      // If a resource is not allowed to have children then it is a page-ish resource
+      // Thus, it can move to within any resource and no further checks are needed
+      if (!isAllowedToHaveChildren(existingResource.type)) return false
+
+      // A resource should not be able to move to its nested children
+      return (
+        nestedChildrenOfExistingResource.some(
+          (child) => child.id === resourceItem.id,
+        ) || false
+      )
+    },
+    [existingResource, nestedChildrenOfExistingResource],
+  )
+
+  const hasParentInStack = useMemo(
+    () =>
+      (resourceStack.length === 1 && !isResourceHighlighted) ||
+      resourceStack.length > 1,
+    [resourceStack.length, isResourceHighlighted],
+  )
+
+  const fullPermalink: string = useMemo(() => {
+    return resourceStack.map((resource) => resource.permalink).join("/")
+  }, [resourceStack])
+
+  const handleClickResourceItem = (
+    resourceItemWithAncestryStack: ResourceItemContent[],
+  ): void => {
+    const lastChild: ResourceItemContent | undefined =
+      lastResourceItemInAncestryStack(resourceItemWithAncestryStack)
+
+    // this should never happen. only added here to satisfy typescript
+    if (!lastChild) return
+
+    const isItemHighlighted = isResourceIdHighlighted(lastChild.id)
+    const canClickIntoItem =
+      lastChild.type === ResourceType.Folder ||
+      lastChild.type === ResourceType.Collection
+
+    if (isItemHighlighted && canClickIntoItem) {
+      setIsResourceHighlighted(false)
+      return
+    }
+
+    setResourceStack(resourceItemWithAncestryStack)
+    setIsResourceHighlighted(true)
+  }
+
+  const handleClickBackButton = useCallback(() => {
+    if (isResourceHighlighted) {
+      setIsResourceHighlighted(false)
+      removeFromStack(2)
+    } else {
+      removeFromStack(1)
+    }
+  }, [isResourceHighlighted, removeFromStack])
+
+  const pendingMovedItemAncestryStack: ResourceItemContent[] =
+    trpc.resource.getAncestryWithSelf
+      .useSuspenseQuery({
+        siteId: String(siteId),
+        resourceId: selectedResourceId,
+      })[0]
+      .map((resource) => ({ ...resource, resourceId: resource.id }))
+
+  useEffect(() => {
```

Comment:
this effect seems redundant. Since this is a suspenseQuery, you can probably init the `useState` directly instead of using the effect (since this effect seems to be run only once)

```ts
const pendingMovedItemAncestryStack = trpc.resource.getAncestryWithSelf
      .useSuspenseQuery(...)

const [resourceStack, setResourceStack] = useState<ResourceItemContent[]>(pendingMovedItemAncestryStack) // probably good to rename this too
```

can move this up if need be.